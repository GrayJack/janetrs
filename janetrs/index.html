<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Janetrs"><title>janetrs - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-47e7ab555ef2818a.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="janetrs" data-themes="" data-resource-suffix="" data-rustdoc-version="1.74.0-nightly (030e4d382 2023-09-10)" data-channel="nightly" data-search-js="search-5d3eaacf19ebf04f.js" data-settings-js="settings-74424d7eec62a23e.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-f194925aa375ae96.css" data-theme-dark-css="dark-1dd4d1ce031e15de.css" data-theme-ayu-css="ayu-49e58d069f567085.css" ><script src="../static.files/storage-db41da1a38ea3cb8.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-8d035c8cea6edbc4.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-f194925aa375ae96.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1dd4d1ce031e15de.css"><link rel="stylesheet" href="../static.files/noscript-cffde32267a19fd6.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../janetrs/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../janetrs/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate janetrs</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 0.6.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Aliases</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">janetrs</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/janetrs/lib.rs.html#1-86">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="janetrs"><a href="#janetrs">Janetrs</a></h2>
<p>A crate with high level bindings to Janet C API.</p>
<h3 id="goals"><a href="#goals">Goals</a></h3>
<p>Provide a safe and ergonomic interface to the Janet C API to create Janet clients and
Janet modules/libraries using Rust.</p>
<p>This project still are in it’s early stages, so breaking changes may happen, there is
no minimal supported Rust version (MSRV) yet.</p>
<p>Notice that most doc tests will fail if the feature “almagation” or “link-system”
aren’t set, because most of then need it for the Janet runtime to function properly.</p>
<h3 id="cargo-features"><a href="#cargo-features">Cargo Features</a></h3>
<ul>
<li><code>std</code>: Enable some trait impl for types that only exist on the <code>std</code> and the Error
trait</li>
<li><code>unicode</code>: Enable more methods for JanetString and JanetBuffer</li>
<li><code>inline-more</code>: More aggressive inlining</li>
<li><code>amalgation</code>: Link the Janet runtime to the package, enabling to use the client
module</li>
<li><code>unicode</code>: Enable some unicode methods for JanetString and JanetBuffer</li>
<li><code>system</code>: Use system header to get Janet functions</li>
<li><code>link-system</code>: Link the Janet runtime to the package from the system, enabling to
use the client module</li>
<li><code>nightly</code>: Enable some parts of the crate that uses nightly features, to use this
feature you must compile the crate using a nightly rust version</li>
</ul>
<h3 id="environment-variables"><a href="#environment-variables">Environment variables</a></h3>
<p><strong>These variables are only used when the <code>amalgation</code> feature is enabled</strong></p>
<p>It is possible to use environment variables to overwrite some Janet definitions.</p>
<ul>
<li><code>JANET_RECURSION_GUARD=&lt;integer&gt;</code></li>
<li><code>JANET_MAX_PROTO_DEPTH=&lt;integer&gt;</code></li>
<li><code>JANET_MAX_MACRO_EXPAND=&lt;integer&gt;</code></li>
<li><code>JANET_STACK_MAX=&lt;integer&gt;</code></li>
</ul>
<h3 id="licensing"><a href="#licensing">Licensing</a></h3>
<p>This software is licensed under the terms of the <a href="./LICENSE">MIT Public License</a>.</p>
<h4 id="todo-typestraits-lacking-or-incomplete"><a href="#todo-typestraits-lacking-or-incomplete">TODO: Types/Traits: Lacking or Incomplete</a></h4>
<ul>
<li><input disabled="" type="checkbox"/>
Marshaling</li>
</ul>
<p><code>[ ]: Lacking</code>
<code>[I]: Incomplete</code>
<code>[X]: Done</code></p>
<p>Probably there is much more missing, for that you can use the
<a href="../evil_janet/index.html" title="mod evil_janet"><code>lowlevel</code></a> module to access the raw C API of Janet if needed</p>
<h4 id="todo-lib-level"><a href="#todo-lib-level">TODO: Lib level</a></h4>
<ul>
<li>Better docs.</li>
<li>Marshalling mechanism</li>
</ul>
</div></details><h2 id="reexports" class="small-section-header"><a href="#reexports">Re-exports</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.lowlevel"><code>pub use <a class="mod" href="../evil_janet/index.html" title="mod evil_janet">evil_janet</a> as lowlevel;</code></div></li></ul><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="allocator/index.html" title="mod janetrs::allocator">allocator</a></div><div class="desc docblock-short">This module provides a allocator that uses the Janet scratch memory API to create
objects tracked by the Janet Garbage Collector.</div></li><li><div class="item-name"><a class="mod" href="array/index.html" title="mod janetrs::array">array</a></div><div class="desc docblock-short">Janet array (vector) type.</div></li><li><div class="item-name"><a class="mod" href="buffer/index.html" title="mod janetrs::buffer">buffer</a></div><div class="desc docblock-short">Janet buffer (string) type.</div></li><li><div class="item-name"><a class="mod" href="client/index.html" title="mod janetrs::client">client</a><span class="stab portability" title="Available on crate features `amalgation` or `link-system` only"><code>amalgation</code> or <code>link-system</code></span></div><div class="desc docblock-short">This module implements anything required to run a Janet client.</div></li><li><div class="item-name"><a class="mod" href="env/index.html" title="mod janetrs::env">env</a></div><div class="desc docblock-short">Module for the Janet environment structure and methods.</div></li><li><div class="item-name"><a class="mod" href="fiber/index.html" title="mod janetrs::fiber">fiber</a></div><div class="desc docblock-short">Janet fibers (soft threads) type.</div></li><li><div class="item-name"><a class="mod" href="function/index.html" title="mod janetrs::function">function</a></div><div class="desc docblock-short">Janet function types.</div></li><li><div class="item-name"><a class="mod" href="io/index.html" title="mod janetrs::io">io</a></div><div class="desc docblock-short">Module for I/O related types and functions</div></li><li><div class="item-name"><a class="mod" href="janet_abstract/index.html" title="mod janetrs::janet_abstract">janet_abstract</a></div><div class="desc docblock-short">Module of the JanetAbstract abstractions.</div></li><li><div class="item-name"><a class="mod" href="pointer/index.html" title="mod janetrs::pointer">pointer</a></div><div class="desc docblock-short">Module for the Janet Pointer type.</div></li><li><div class="item-name"><a class="mod" href="string/index.html" title="mod janetrs::string">string</a></div><div class="desc docblock-short">Janet String type.</div></li><li><div class="item-name"><a class="mod" href="structs/index.html" title="mod janetrs::structs">structs</a></div><div class="desc docblock-short">Janet Struct (immutable HashMap) type.</div></li><li><div class="item-name"><a class="mod" href="symbol/index.html" title="mod janetrs::symbol">symbol</a></div><div class="desc docblock-short">Janet symbols and keywords types.</div></li><li><div class="item-name"><a class="mod" href="table/index.html" title="mod janetrs::table">table</a></div><div class="desc docblock-short">Janet table (mutable HashMap) type.</div></li><li><div class="item-name"><a class="mod" href="tuple/index.html" title="mod janetrs::tuple">tuple</a></div><div class="desc docblock-short">Janet tuple type.</div></li><li><div class="item-name"><a class="mod" href="util/index.html" title="mod janetrs::util">util</a></div><div class="desc docblock-short">Module for stuff that are not required either to use in a application or to write
janet modules.</div></li></ul><h2 id="macros" class="small-section-header"><a href="#macros">Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.array.html" title="macro janetrs::array">array</a></div><div class="desc docblock-short">Creates a <a href="./types/tuple/struct.JanetArray.html"><code>JanetArray</code></a> containing the arguments.</div></li><li><div class="item-name"><a class="macro" href="macro.bad_slot.html" title="macro janetrs::bad_slot">bad_slot</a></div><div class="desc docblock-short">A macro helper to use the default message when getting the wrong types in the function
argument when the situations that are more complex than the ones handled in
<a href="trait.JanetArgs.html" title="trait janetrs::JanetArgs"><code>JanetArgs</code></a>, like multiple accepted types.</div></li><li><div class="item-name"><a class="macro" href="macro.check_janet_version.html" title="macro janetrs::check_janet_version">check_janet_version</a></div><div class="desc docblock-short">Checks the specified version of Janet is equal of the used Janet. Emits a boolean.</div></li><li><div class="item-name"><a class="macro" href="macro.declare_janet_mod.html" title="macro janetrs::declare_janet_mod">declare_janet_mod</a></div><div class="desc docblock-short">Declare a Janet module.</div></li><li><div class="item-name"><a class="macro" href="macro.janet_mod.html" title="macro janetrs::janet_mod">janet_mod</a><span class="stab deprecated" title="">Deprecated</span></div><div class="desc docblock-short">A macro to make life easier creating modules for Janet from Rust.</div></li><li><div class="item-name"><a class="macro" href="macro.jcatch.html" title="macro janetrs::jcatch">jcatch</a></div><div class="desc docblock-short">Execute a function (<a href="./types/struct.JanetCFunction.html"><code>JanetCFunction</code></a>, Rust function or extern C function) and catch
any janet panic that may happen as a <a href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result"><code>Result</code></a>.</div></li><li><div class="item-name"><a class="macro" href="macro.jpanic.html" title="macro janetrs::jpanic">jpanic</a></div><div class="desc docblock-short">Causes a panic in the Janet side (exception). Diferently of the Rust <code>panic!</code> macro,
this macro does <strong>not</strong> terminate the current thread. Instead, it sends a error signal
with the passed message where the Janet runtime takes care to properly exit.</div></li><li><div class="item-name"><a class="macro" href="macro.jtry.html" title="macro janetrs::jtry">jtry</a><span class="stab portability" title="Available on crate feature `std` only"><code>std</code></span></div><div class="desc docblock-short">Macro that tries to run a expression, and if it panics in the Rust side, it tries to
recover from that and pass the Rust panic string to a Janet Panic.</div></li><li><div class="item-name"><a class="macro" href="macro.structs.html" title="macro janetrs::structs">structs</a></div><div class="desc docblock-short">Creates a <a href="./types/tuple/struct.JanetStruct.html"><code>JanetStruct</code></a> containing the arguments key-value pairs.</div></li><li><div class="item-name"><a class="macro" href="macro.table.html" title="macro janetrs::table">table</a></div><div class="desc docblock-short">Creates a <a href="./types/tuple/struct.JanetTable.html"><code>JanetTable</code></a> containing the arguments key-value pairs.</div></li><li><div class="item-name"><a class="macro" href="macro.tuple.html" title="macro janetrs::tuple">tuple</a></div><div class="desc docblock-short">Creates a <a href="./types/tuple/struct.JanetTuple.html"><code>JanetTuple</code></a> containing the arguments.</div></li></ul><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Janet.html" title="struct janetrs::Janet">Janet</a></div><div class="desc docblock-short"><code>Janet</code> is the central structure of the Janet Language.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetAbstract.html" title="struct janetrs::JanetAbstract">JanetAbstract</a></div><div class="desc docblock-short">Type that represents the Janet Abstract type.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetArray.html" title="struct janetrs::JanetArray">JanetArray</a></div><div class="desc docblock-short">Janet <a href="https://janet-lang.org/docs/data_structures/arrays.html">arrays</a> are a fundamental
datatype in Janet. Janet Arrays are values that contain a sequence of other values.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetBuffer.html" title="struct janetrs::JanetBuffer">JanetBuffer</a></div><div class="desc docblock-short">Janet <a href="https://janet-lang.org/docs/data_structures/buffers.html">buffers</a> are the mutable
version of <a href="./../string/struct.JanetString.html"><code>JanetStrings</code></a>. Since Janet strings can hold any sequence of bytes,
including zeros, buffers share this same property and can be used to hold any
arbitrary memory, which makes them very simple but versatile data structures. They can
be used to accumulate small strings into a large string, to implement a bitset, or to
represent sound or images in a program.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetConversionError.html" title="struct janetrs::JanetConversionError">JanetConversionError</a></div><div class="desc docblock-short">The error when converting <a href="struct.Janet.html" title="struct janetrs::Janet"><code>Janet</code></a>s to C Janet types.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetFiber.html" title="struct janetrs::JanetFiber">JanetFiber</a></div><div class="desc docblock-short">A lightweight green thread in Janet. It does not correspond to operating system
threads.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetFunction.html" title="struct janetrs::JanetFunction">JanetFunction</a></div><div class="desc docblock-short">A representation of a Janet function defined at the Janet side.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetGc.html" title="struct janetrs::JanetGc">JanetGc</a></div><div class="desc docblock-short">The Janet Garbage Collector type.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetGcLockGuard.html" title="struct janetrs::JanetGcLockGuard">JanetGcLockGuard</a></div><div class="desc docblock-short">An RAII implementation of a “scoped lock” of the Janet GC. When this structure is
dropped (falls out of scope), the lock will be unlocked.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetGcRootGuard.html" title="struct janetrs::JanetGcRootGuard">JanetGcRootGuard</a></div><div class="desc docblock-short">An RAII implementation of a rooted <a href="struct.Janet.html" title="struct janetrs::Janet"><code>Janet</code></a> value. When this structure is dropped
(falls out of scope), the rooting will be undone.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetKeyword.html" title="struct janetrs::JanetKeyword">JanetKeyword</a></div><div class="desc docblock-short">Janet keyword. Janet being a lisp-like language a keyword is not a especial word of
the language, it is a normal string that can be defined by the user.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetPointer.html" title="struct janetrs::JanetPointer">JanetPointer</a></div><div class="desc docblock-short">Janet pointer type.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetRng.html" title="struct janetrs::JanetRng">JanetRng</a></div><div class="desc docblock-short">Type representing the Janet pseudorandom number generator</div></li><li><div class="item-name"><a class="struct" href="struct.JanetString.html" title="struct janetrs::JanetString">JanetString</a></div><div class="desc docblock-short">Janet strings are a immutable type that are similar to <a href="./../buffer/struct.JanetBuffer.html">Janet buffers</a>.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetStruct.html" title="struct janetrs::JanetStruct">JanetStruct</a></div><div class="desc docblock-short"><code>JanetStruct</code>s are immutable data structures that map keys to values.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetSymbol.html" title="struct janetrs::JanetSymbol">JanetSymbol</a></div><div class="desc docblock-short">Janet symbol type. Usually used to name things in Janet.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetTable.html" title="struct janetrs::JanetTable">JanetTable</a></div><div class="desc docblock-short">Janet <a href="https://janet-lang.org/docs/data_structures/tables.html">table</a>s are mutable data structures that map keys to values. Values are put
into a Janet table with a key, and can be looked up later with the same key. Tables
are implemented with an underlying open hash table, so they are quite fast and cache
friendly.</div></li><li><div class="item-name"><a class="struct" href="struct.JanetTuple.html" title="struct janetrs::JanetTuple">JanetTuple</a></div><div class="desc docblock-short">Janet <a href="https://janet-lang.org/docs/data_structures/tuples.html">tuples</a> are immutable,
sequential types that are similar to <a href="array/struct.JanetArray.html" title="struct janetrs::array::JanetArray">Janet arrays</a>.</div></li></ul><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.JanetSignal.html" title="enum janetrs::JanetSignal">JanetSignal</a></div><div class="desc docblock-short">Signals that can be produced by a <a href="function/struct.JanetFunction.html" title="struct janetrs::function::JanetFunction"><code>JanetFunction</code></a> and <a href="function/type.JanetCFunction.html" title="type janetrs::function::JanetCFunction"><code>JanetCFunction</code></a>
representing that those worked correctly or not.</div></li><li><div class="item-name"><a class="enum" href="enum.JanetType.html" title="enum janetrs::JanetType">JanetType</a></div><div class="desc docblock-short">Representation of all Janet types.</div></li><li><div class="item-name"><a class="enum" href="enum.TaggedJanet.html" title="enum janetrs::TaggedJanet">TaggedJanet</a></div><div class="desc docblock-short">Janet type in the form of a Tagged Union.</div></li></ul><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.DeepEq.html" title="trait janetrs::DeepEq">DeepEq</a></div><div class="desc docblock-short">A trait to express a deep equality.</div></li><li><div class="item-name"><a class="trait" href="trait.IsJanetAbstract.html" title="trait janetrs::IsJanetAbstract">IsJanetAbstract</a></div><div class="desc docblock-short">The trait that encodes the information required to instatiate the implementer as
<a href="janet_abstract/struct.JanetAbstract.html" title="struct janetrs::janet_abstract::JanetAbstract"><code>JanetAbstract</code></a></div></li><li><div class="item-name"><a class="trait" href="trait.JanetArgs.html" title="trait janetrs::JanetArgs">JanetArgs</a></div><div class="desc docblock-short">Trait that only exist to extend methods over <code>[Janet]</code> so it’s easier to get
<a href="attr.janet_fn.html" title="attr janetrs::janet_fn"><code>janet_fn</code></a> args.</div></li><li><div class="item-name"><a class="trait" href="trait.JanetExtend.html" title="trait janetrs::JanetExtend">JanetExtend</a></div><div class="desc docblock-short">Trait that express the ability of a Janet collection to extend it with another
collection.</div></li><li><div class="item-name"><a class="trait" href="trait.JanetTypeName.html" title="trait janetrs::JanetTypeName">JanetTypeName</a></div><div class="desc docblock-short">Trait defining the name of the type known to Janet</div></li></ul><h2 id="types" class="small-section-header"><a href="#types">Type Aliases</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.JanetCFunction.html" title="type janetrs::JanetCFunction">JanetCFunction</a></div><div class="desc docblock-short">C function pointer that is accepted by Janet as a type.</div></li></ul><h2 id="attributes" class="small-section-header"><a href="#attributes">Attribute Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.cjvg.html" title="attr janetrs::cjvg">cjvg</a></div><div class="desc docblock-short">Conditional Janet Version Gate</div></li><li><div class="item-name"><a class="attr" href="attr.janet_fn.html" title="attr janetrs::janet_fn">janet_fn</a></div><div class="desc docblock-short">Macro that tranforms a high-level Janet function (<code>fn(&amp;mut [Janet]) -&gt; Janet</code>)
to the thing the Janet C API is expecting (<code>fn(i32, *mut janetrs::lowlevel::Janet) -&gt; janetrs::lowlevel::Janet</code>)</div></li><li><div class="item-name"><a class="attr" href="attr.janet_version.html" title="attr janetrs::janet_version">janet_version</a></div><div class="desc docblock-short">Conditional Janet Version Gate</div></li></ul></section></div></main></body></html>