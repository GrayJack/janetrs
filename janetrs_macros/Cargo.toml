[package]
name = "janetrs_macros"
version = "0.4.6"
authors = ["Eric Shimizu Karbstein <gr41.j4ck@gmail.com>"]
description = "Attribute macros for JanetRS"
repository = "https://github.com/GrayJack/janetrs"
homepage = "https://crates.io/crates/janetrs_macros"
documentation = "https://docs.rs/janetrs_macros"
license-file = "../LICENSE"
edition = "2021"
rust-version = "1.56.0"
keywords = ["janet", "embedding", "proc-macro"]
categories = ["development-tools::procedural-macro-helpers"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
name = "janetrs_macros"
path = "src/lib.rs"
proc-macro = true

[dependencies]
syn = { version = "1", features = ["extra-traits", "full"] }
quote = "1"
proc-macro2 = "1"
janetrs_version = "0.1"

[features]
default = []
system = ["janetrs_version/system"]

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0", features = ["diff"] }
janetrs = { path = "../../janetrs" }
