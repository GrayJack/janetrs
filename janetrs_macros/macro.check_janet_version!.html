<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.check_janet_version.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.check_janet_version.html">macro.check_janet_version.html</a>...</p>
    <script>location.replace("macro.check_janet_version.html" + location.search + location.hash);</script>
</body>
</html>