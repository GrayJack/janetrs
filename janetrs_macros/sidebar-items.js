window.SIDEBAR_ITEMS = {"attr":["cjvg","janet_fn","janet_version"],"macro":["check_janet_version","declare_janet_mod"]};