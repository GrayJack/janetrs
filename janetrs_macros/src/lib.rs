extern crate proc_macro;

use quote::{quote, quote_spanned};
use syn::{parse_macro_input, spanned::Spanned};

#[proc_macro_attribute]
pub fn janet_fn(
    _args: proc_macro::TokenStream, input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let func = parse_macro_input!(input as syn::Item);

    let ts = if let syn::Item::Fn(f) = func {
        // dbg!(&f);
        let f_clone = f.clone();
        let attrs = f.attrs;
        let vis = f.vis;
        let name = f.sig.ident;
        // let syn::ReturnType::Type(_, type_path) = f.sig.output;

        // check for inputs
        if matches!(f.sig.inputs.len(), 1) {
            if let Some(syn::FnArg::Typed(syn::PatType { ty, .. })) = f.sig.inputs.first() {
                if let syn::Type::Reference(syn::TypeReference {
                    mutability,
                    ref elem,
                    ..
                }) = **ty
                {
                    if mutability.is_none() {
                        return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found a immutable reference");}.into();
                    } else if let syn::Type::Slice(syn::TypeSlice {
                        elem: ref slice, ..
                    }) = **elem
                    {
                        if let syn::Type::Path(syn::TypePath { ref path, .. }) = **slice {
                            if !janet_path_checker(path) {
                                return quote_spanned! {path.span() => compile_error!("expected to be a `janetrs::types::Janet`");}.into();
                            }
                        } else {
                            return quote_spanned! {slice.span() => compile_error!("expected to be a `janetrs::types::Janet`");}.into();
                        }
                    } else {
                        return quote_spanned! {elem.span() => compile_error!("expected to be a slice of `janetrs::types::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {ty.span() => compile_error!("expected argument to be a mutable reference and found something that is not a reference at all");}.into();
                }
            }

            // check output type

            let output_span = match f.sig.output.clone() {
                syn::ReturnType::Default => f_clone.sig.span(),
                syn::ReturnType::Type(_, ty) => ty.span(),
            };
            if let syn::ReturnType::Type(_, ty) = f.sig.output {
                if let syn::Type::Path(syn::TypePath { ref path, .. }) = *ty {
                    if !janet_path_checker(path) {
                        return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
                    }
                } else {
                    return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
                }
            } else {
                return quote_spanned! {output_span => compile_error!("expected return type to be `janetrs::types::Janet`");}.into();
            }

            quote! {
                #(#attrs)*
                #[no_mangle]
                #vis unsafe extern "C" fn #name(argc: i32, argv: *mut ::janetrs::lowlevel::Janet) -> ::janetrs::lowlevel::Janet {
                    #f_clone

                    let args = unsafe { core::slice::from_raw_parts_mut(argv, argc as usize) };
                    let mut args = unsafe { &mut *(args as *mut [::janetrs::lowlevel::Janet] as *mut [::janetrs::types::Janet])};
                    #name(args).into()
                }
            }
        } else {
            quote_spanned! {f_clone.sig.inputs.span() => compile_error!("expected exactly one argument of type `&mut [janetrs::types::Janet]`");}
        }
    } else {
        quote_spanned! {func.span() => compile_error!("expected fn item");}
    };

    ts.into()
}

fn janet_path_checker(path: &syn::Path) -> bool {
    match path.segments.len() {
        1 => {
            let ident = if let Some(i) = path.get_ident() {
                i
            } else {
                return false;
            };
            let test = syn::Ident::new("Janet", ident.span());

            *ident == test
        },
        2 => {
            let mod_types = &path.segments.first().unwrap().ident;
            let janet_ident = &path.segments.last().unwrap().ident;

            let mod_expected = syn::Ident::new("types", mod_types.span());
            let janet_expected = syn::Ident::new("Janet", janet_ident.span());

            *mod_types == mod_expected && *janet_ident == janet_expected
        },
        3 => {
            let mut iter = path.segments.iter();
            let (janetrs_mod, types_mod, janet_ident) = (
                &iter.next().unwrap().ident,
                &iter.next().unwrap().ident,
                &iter.next().unwrap().ident,
            );

            let janetrs_expected = syn::Ident::new("janetrs", janetrs_mod.span());
            let types_expected = syn::Ident::new("types", types_mod.span());
            let janet_expected = syn::Ident::new("Janet", janet_ident.span());

            *janetrs_mod == janetrs_expected
                && *types_mod == types_expected
                && *janet_ident == janet_expected
        },
        _ => false,
    }
}
