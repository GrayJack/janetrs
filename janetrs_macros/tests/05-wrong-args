#![allow(unused_imports)]

use janetrs_macros::*;
use janetrs::Janet;

#[janet_fn]
pub fn function(_args: &mut [Janet], _wrong: Janet) -> Janet {
    Janet::nil()
}

fn main() {}