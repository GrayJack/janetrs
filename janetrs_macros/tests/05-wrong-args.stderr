error: expected exactly one argument of type `&mut [janetrs::types::Janet]`
 --> $DIR/05-wrong-args:7:17
  |
7 | pub fn function(_args: &mut [Janet], _wrong: Janet) -> Janet {
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
